Add new Statistics Component

kahrendt opened the pull request:
# What does this implement/fix?

This component quickly computes nine summary statistics like mean, standard deviation, and trend from a source sensor's measurements. The component computes the statistics over a sliding window or a resettable continuous window. The implemented algorithms compute statistics extremely efficiently (constant time insertion, eviction, and retrieval while only storing n+2 summary statistics for n measurements in a sliding window). Further, I have attempted to ensure that all operations are as numerically stable as possible, especially for continuous windows with measurements spanning days/weeks/months.

This component has a lot of code, but I tried to document everything as thoroughly as possible. The header of statistics.h describes the layout of the code in detail. In particular, it defines the terms I use throughout the code and summarizes the purpose of each header file. Each specific header file further describes its purpose, explains the memory usage and computational complexity (if applicable), and discusses some of the mathematics used with references to all papers consulted. I documented every method. I added comments directly to the code when necessary to further improve clarity. Please ask any questions you have, as I have been working on this for a month and a half straight, so it is quite possible that I didn't document something fully!

I would appreciate anyone trying this out and letting me know how things work. I have attempted to test many cases, but with a component this complicated, I could have easily missed something. I would also appreciate ideas on better setups for the configuration or anything else that would improve the code and its compatibility with ESPHome.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes esphome/feature-requests#2185, fixes esphome/feature-requests#2295,

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3067

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml

  - platform: statistics
    source_id: template_sensor
    window:
      type: sliding
      window_size: 10
      send_every: 10
      send_first_at: 10
    statistics:
      - name: "Template count"  # should be 10
        type: count
      - name: "Template duration" # should be around 10s
        type: duration
      - name: "Template max" # should be 18.0
        type: max
      - name: "Template min" # should be 1.0
        type: min
      - name: "Template mean" # should be 7.8
        type: mean
      - name: "Template argmax" # should be around 1s
        type: since_argmax
      - name: "Template argmin" # should be around 4s
        type: since_argmin
      - name: "Template std_dev" # should be around 5.750
        type: std_dev
      - name: "Template Trend" # should be around 0.994
        type: trend

  - platform: template
    name: "template sensor"
    id: template_sensor
    update_interval: 1s
    state_class: measurement
    lambda: |-
      // template sensor loops through each value in the array every 10 seconds
      // with this known set of measurements, all stats can be manually computed for verification
      static int array[] = {2,3,9,5,9,1,13,14,18,4};
      static int index = 9;

      if (index == 9)
        index = 0;
      else
        ++index;

      return array[index];

      if (index == 8)
        index = 0;
      else
        ++index;

      return array[index];
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

kahrendt committed:
initial commit for pull request

SeByDocKy commented:
Hi great adding. I had the same idea some months ago: https://github.com/SeByDocKy/myESPhome/tree/main/components/statistics

But your version is more elaborate. Are  all statistics computed "on-line" (just need to store x_n and eventually statistics_{n-1}) ?

kahrendt commented:
> Hi great adding. I had the same idea some months ago: https://github.com/SeByDocKy/myESPhome/tree/main/components/statistics
> 
> But your version is more elaborate. Are all statistics computed "on-line" (just need to store x_n and eventually statistics_{n-1}) ? or need a more deeper buffer of deph n ?

Interesting, there is a lot of overlap with your component and mine!

If you use the ``continuous`` window type, then the component is computed online as you mentioned, much like your component. The ``chunked_continuous`` type stores log(n) statistics for n measurements; this is to maintain numerical stability with very large windows.

The ``sliding`` and ``chunked_sliding`` window types use a buffer of length n where new measurements are added or older measurements are removed. In fact, the ``sliding`` type should behave identically to the ``sliding_window_moving_average`` sensor filter (though with a lot more statistics available).

kahrendt commented:
I'm changing this back to a draft as I found a bug with argmin and argmax not behaving when the sensor state is restored. I have a fix in the works to store the Unix time for each instead of the time since boot.

kahrendt committed:
Simplify codegen+configuration logic

kahrendt committed:
use set_interval for durations

kahrendt commented:
The current version has solved the bug with the argmin and argmax sensors being nonsensical when restoring from flash. I am keeping the PR as a draft while I work on other improvements. I am changing chunk duration handling to use the scheduler's set_interval method instead of checking duration only on a new sensor measurement. This change makes the timing much more accurate and simplifies the code logic.

kahrendt committed:
add automation trigger

kahrendt committed:
remove coeffecient of determination computation

kahrendt committed:
fix floating types for queues

kahrendt commented:
I have reopened this for review. 

- Fixed bug with ``since_*`` sensors after restoring the state from flash on reboot.
- The ``duration`` and ``since_*`` sensors use the configured ``time_unit``. For durations, floating point precision quickly becomes an issue when reporting the time in the natively stored integer milliseconds.
- I have simplified/unified the window configuration and the codegen logic.
- Simplified the logic in statistics.cpp with respect to handling overflows and disabling certain continuous window options.
- I have changed the ``chunk_duration`` setting to use ESPHome's scheduler, so time based windows are much more precise in duration. 
- I have added an ``on_update`` trigger that runs automations after all sensors have safely updated. That trigger gives the raw ``Aggregate`` object as the variable ``x``, and I have added to the documentation on how it can be used in lambdas.

kahrendt committed:
add lambda sensor type

kahrendt committed:
pass last sensor measurement to lambdas

kahrendt committed:
clang fixes

kahrendt committed:
improve documentation

kahrendt committed:
dump_config LogString helper functions are static

kahrendt commented:
This PR is ready for review. I have been using this for the last few months in several contexts, and I have made changes that improve the component. First, Aggregates store the group type and weight type rather than passing it when calling various functions. Second, I moved more logic into the Python codegen. The Python code determines which raw statistics are necessary to compute the configured statistics sensors. Also, the codegen provides a lambda call that returns the desired statistic. Finally, I added all statistic sensors to a vector. Now, users can configure more than one sensor of each type, which is especially useful for multiple lambda sensors. Lambda sensors can use the current Aggregate and the most recent source sensor measurement. Now, lambda sensors can compute z-scores.

I had to introduce some helper functions in ``sensor.py``. The current inherit properties helper does not work well with a typed sensors list. The helpers I added make it easier to inherit properties from the source sensor in this scenario.

kevdel commented:
any chance of this being available for the 8266 platform?

pos-ei-don commented:
Can i already use this for a Test in my esphome? just including it as a package?

kahrendt commented:
I haven't worked on this in quite awhile (somehow I accidently force-pushed the branch recently though), but I have had it running in production for a long time across many devices. The current form of the PR is way too big to be reviewable, so when I revisit this in the future, I will try to split it up more or move some of the algorithms into their own library.

> any chance of this being available for the 8266 platform?

It should work on it already, just be careful to not use too large of a sliding window or too many statistic sensors in general, as there isn't a lot of memory. I use it on all of my converted smart plugs.

> Can i already use this for a Test in my esphome? just including it as a package?

You should be able to add it as an external component without issue.

pos-ei-don commented:
> You should be able to add it as an external component without issue.

i tried, but stuggled. Can you tell me what URL you set for the component?
Thank you a lot!!

kevdel commented:
> > You should be able to add it as an external component without issue.
> 
> i tried, but stuggled. Can you tell me what URL you set for the component? Thank you a lot!!

try this

```
external_components:
  - source: github://pr#5103
    components: [ statistics ]
```

pos-ei-don commented:
Thank you a lot! Works great!!!!

kevdel commented:
issue on compliation
ESP8266 board

using the following config

```
  - platform: statistics
    source_id: sens_1
    window:
      type: sliding
      window_size: 10
      send_every: 5
      send_first_at: 5
```


```
In file included from src/esphome/components/sensor/sensor.h:3,
                 from src/esphome/components/statistics/statistics.h:74,
                 from src/esphome/components/statistics/statistics.cpp:1:
src/esphome/components/statistics/statistics.cpp: In member function 'virtual void esphome::statistics::StatisticsComponent::dump_config()':
src/esphome/core/log.h:173:25: error: initializer fails to determine size of '__pstr__'
  173 | #define LOG_STR_ARG(s) ((PGM_P) (s))
      |                         ^
src/esphome/core/log.h:97:90: note: in definition of macro 'esph_log_config'
   97 |   esp_log_printf_(ESPHOME_LOG_LEVEL_CONFIG, tag, __LINE__, ESPHOME_LOG_FORMAT(format), ##__VA_ARGS__)

```

kevdel commented:
Was able to get it to compile successfully by changing the level of the logger component

Results of investigations - checked means it compiled. Unchecked results in the failure noted above

- [x] None
- [x] Warn
- [x] Error
- [x] Info
- [x] Debug
- [ ] Verbose
- [ ] Very Verbose


```
logger:
  level: NONE
```

kevdel commented:
Another issue found on an esp8266 yaml

the on_value_range event does not fire

```      name: "Temperature Trend"
      id: temp_trend_1
      unit_of_measurement: "F"
      type: trend  
      on_value:
       then:
        - logger.log:
            format: "this log message is visible"
            level: INFO
      on_value_range: # between the rising trend line
      - above: 0.03
        below: 0.4
        then:
        - logger.log:
            format: "this log message is NOT visible"
            level: INFO
```

cinadr commented:
Hi @kahrendt!

Successfully using this for my pool heater on ESP32 for a year or so. Can you add support for **web_server sorting groups** or **entity_category**? 

Thank you for your work!
